compiler/hash-typecheck/src/ops/unify.rs

```diff
@@ -170,15 +171,62 @@ impl<'gs, 'ls, 'cd, 's> Unifier<'gs, 'ls, 'cd, 's> {
         Ok(sub)
     }
 
+    /// Unify the two given argument lists, by argument-wise unifying terms.
+    /// The function requires a reference to the parent source and target
+    /// terms in order to give meaningful error messages.
+    pub(crate) fn unify_args(
+        &mut self,
+        src_args_id: ArgsId,
+        target_args_id: ArgsId,
+        src_id: TermId,
+        target_id: TermId,
+    ) -> TcResult<Sub> {
+        let src_args = self.args_store().get(src_args_id).clone();
+        let target_args = self.args_store().get(target_args_id).clone();
+
+        let cannot_unify = |reason: ParamUnificationErrorReason| {
+            Err(TcError::CannotUnifyArgs {
+                src_args_id,
+                target_args_id,
+                reason,
+                src: src_id,
+                target: target_id,
+            })
+        };
+
+        // Ensure the argument lengths match
+        if src_args.positional().len() != target_args.positional().len() {
+            return cannot_unify(ParamUnificationErrorReason::LengthMismatch);
+        }
+
+        // For each argument, ensure it is the same:
+        let mut cumulative_sub = Sub::empty();
+        let pairs = src_args.positional().iter().zip(target_args.positional());
+        for (index, (src_param, target_param)) in pairs.enumerate() {
+            // Names match
+            if src_param.name != target_param.name {
+                return cannot_unify(ParamUnificationErrorReason::NameMismatch(index));
+            }
+            // Values match
+            let ty_sub = self.unify_terms(src_param.value, target_param.value)?;
+
+            // Add to cumulative substitution
+            cumulative_sub.extend(&ty_sub);
+        }
```

Comment:
there is a function to perform this kind of validation I think?

Reply:
Which one?

Reply:
`pair_args_with_params` in `compiler/hash-typecheck/src/ops/params.rs` is pretty similar, wonder if can unify all of this logic

Reply:
It's not really the same since the one you said does not perform any unification, just matching. Where as this one performs no matching, just unification.